### PR TITLE
Add rule prefer-type-guard

### LINKED ISSUE #270: Rule : Use Typeguard return type
Implement [S4322](https://jira.sonarsource.com/browse/RSPEC-4322)

```
function isSomething(x: BaseType) : boolean {
  return !!(x as Something).markerProperty;
}
```
Can become
```
function isSomething(x: BaseType) : x is Something {
  return !!(x as Something).markerProperty;
}
```

### PR DESCRIPTION
**Pull request by andrea-guarino-sonarsource**
Fixes #270

### CODE REVIEW

File: sonarts-core/tests/rules/preferTypeGuardRule/preferTypeGuardRule.lint.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
+  function isFish(animal: Animal) {
+//^^^^^^^^ {{Change this boolean return type into the type predicate "animal is Fish".}}
+  return (animal as Fish).swim !== undefined;
+}
+
+function isFish(animal: Animal): animal is Fish {
+  return (animal as Fish).swim !== undefined;
+}
+
+  function isFish(animal: Animal) {
```
Comment (vilchik-elena) [inline sonarts-core/tests/rules/preferTypeGuardRule/preferTypeGuardRule.lint.ts]: what's the difference with the first case?

File: sonarts-core/tests/rules/preferTypeGuardRule/preferTypeGuardRule.lint.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+function isFish(animal: Animal) {
+  return !!animal.name;
+}
+
+  function isFish(animal: Animal) {
+//^^^^^^^^ {{Change this boolean return type into the type predicate "animal is Fish".}}
+  return (<Fish>animal).swim !== undefined;
+}
+
+// OK, to avoid FP at line 48
```
Comment (vilchik-elena) [inline sonarts-core/tests/rules/preferTypeGuardRule/preferTypeGuardRule.lint.ts]: improve this comment

File: sonarts-core/tests/rules/preferTypeGuardRule/preferTypeGuardRule.lint.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+let typePredicate = (animal: Animal) => !!(animal as Fish).swim;
+
+parsedPatterns.filter(parsedPattern => !!(<ParsedStringPattern>parsedPattern).basenames);
```
Comment (vilchik-elena) [inline sonarts-core/tests/rules/preferTypeGuardRule/preferTypeGuardRule.lint.ts]: normalize this code sample

File: sonarts-core/tests/rules/preferTypeGuardRule/preferTypeGuardRule.lint.ts (hunk 1 above)
Comment (andrea-guarino-sonarsource) [reply sonarts-core/tests/rules/preferTypeGuardRule/preferTypeGuardRule.lint.ts]: Here I'm testing `(animal as Fish).swim != undefined;` while in the first case is `return (animal as Fish).swim !== undefined;` (i.e. strict inequality)

Comment (vilchik-elena) [reply sonarts-core/tests/rules/preferTypeGuardRule/preferTypeGuardRule.lint.ts]: indeed:) that's why it's nicer to have descriptive test cases (with comment above or meaningful function name)

Comment (andrea-guarino-sonarsource) [reply sonarts-core/tests/rules/preferTypeGuardRule/preferTypeGuardRule.lint.ts]: Point taken, I will add a comment!

File: sonarts-core/src/rules/preferTypeGuardRule.ts [hunk 4]
```diff
@@ -37,50 +45,55 @@ export class Rule extends tslint.Rules.AbstractRule {
...
     } else if (this.isNegation(returnExpression) && this.isNegation(returnExpression.operand)) {
-      const castedType = this.getCastedTypeFromPropertyAccess(returnExpression.operand.operand);
-      if (castedType) {
-        this.addIssue(functionLikeMainToken(node), Rule.MESSAGE(parameters[0].name.getText(), castedType));
-      }
+      this.checkCastedType(node, parameters.length, returnExpression.operand.operand);
+    }
+  }
+
+  private checkCastedType(node: ts.FunctionLikeDeclaration, nOfParam: number, expression: ts.Expression) {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/preferTypeGuardRule.ts]: you don't need second parameter, you can infer it from first one.

**Pull request merged**